### PR TITLE
Basic wrapper for Calcium qqbar

### PR DESCRIPTION
**Pull request by fredrik-johansson**
This still needs more work (functions, docs, tests); just creating a PR so people can look and comment.

### CODE REVIEW

File: src/calcium/qqbar.jl [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+function numerator(a::fmpq_poly)
+   z = fmpz_poly()
+   ccall((:fmpq_poly_get_numerator, libflint), Nothing,
+         (Ref{fmpz_poly}, Ref{fmpq_poly}), z, a)
+   return z
+end
+
+function roots(f::fmpq_poly, R::CalciumQQBarField)
+    return roots(numerator(f), R)
+end
```
Comment (thofma) [inline src/calcium/qqbar.jl:335]:
```suggestion
function _numerator(a::fmpq_poly)
   z = fmpz_poly()
   ccall((:fmpq_poly_get_numerator, libflint), Nothing,
         (Ref{fmpz_poly}, Ref{fmpq_poly}), z, a)
   return z
end

function roots(f::fmpq_poly, R::CalciumQQBarField)
    return roots(_numerator(f), R)
end
```
This function should not be user facing.

**Comment by thofma** [comment]
Nice. I would be in favor of just calling it `QQBar` instead of `CalciumQQBar`.

File: src/calcium/qqbar.jl (hunk 1 above)
Comment (fredrik-johansson) [reply src/calcium/qqbar.jl:335]: Actually, I thought I should move it to fmpq_poly.jl and document it! We already have denominator() for fmpq_poly.

Comment (thofma) [reply src/calcium/qqbar.jl:335]: This is an illegal polynomial because it has no parents. Users should not be able to create it easily.

**Comment by fredrik-johansson** [comment]
> Nice. I would be in favor of just calling it `QQBar` instead of `CalciumQQBar`.

Yeah, I thought I'd add that as an alias. However, I'm still undecided about the overall naming.

When we have fields of algebraic numbers based on ca_t, it needs to be somehow obvious that users should create their PolynomialRings over those and not over this CalciumQQBar if they want performance. We'll have something like this:

    # slow default field
    QQBar(3)

    # fast ca_t field for calculation
    K = RealAlgebraicField(<optional options>)
    K(3)

    # fast ca_t field for calculation
    K = ComplexAlgebraicField(<optional options>)
    K(3)

Naming suggestions?

*    MinpolyQQBar
*    SlowQQBar
*    FixedQQBar
*    ConstantQQBar
*    Algebraics
*    AlgebraicConstants
*    CanonicalAlgebraicNumbers
*    AlgebraicNumbers

...

**Comment by wbhart** [comment]
Looks ok so far, codecov warnings notwithstanding.

**Comment by fredrik-johansson** [comment]
This is basically done except for writing documentation. Any things in the interface or implementation that need to change?

Things I will not do in this PR:

* Misc conversions (Julia complex numbers, etc)
* Different printing format(?)
* Symbolic expression input/output, serialization
* Hashing
* Polynomial evaluation
* Conversions to/from Antic

**Comment by wbhart** [comment]
Someone will have to implement hash and expressify before this can be merged. But perhaps we can help out there. The other things you don't wish to add to this PR are fine by me.

I will take a look at the rest now.

**Comment by wbhart** [comment]
Unfortunately rand has to be done using the new rand interface as well before we can merge.

File: src/calcium/qqbar.jl [hunk 2]
```diff
@@ -0,0 +1,837 @@
...
+# function Base.hash(a::qqbar, h::UInt)
+# end
+
+###############################################################################
+#
+#   Constructors
+#
+###############################################################################
+
+function qqbar(a::Int)
```
Comment (wbhart) [inline src/calcium/qqbar.jl:46]: Any reason to not make these inner constructors (inside the type definition)?

File: test/calcium/qqbar-test.jl [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+   @test qqbar(2) < 3
+   @test qqbar(2) < fmpz(3)
+   @test qqbar(2) < fmpq(3)
+   @test 2 < qqbar(3)
+   @test fmpz(2) < qqbar(3)
+   @test fmpq(2) < qqbar(3)
+
+end
+
+@testset "qqbar.rand" begin
```
Comment (wbhart) [inline test/calcium/qqbar-test.jl:208]: should call test_rand

File: src/calcium/qqbar.jl [hunk 4]
```diff
@@ -0,0 +1,837 @@
...
+   return z
+end
+
+###############################################################################
+#
+#   Unsafe functions
+#
+###############################################################################
+
+function zero!(z::qqbar)
```
Comment (wbhart) [inline src/calcium/qqbar.jl:799]: We are starting to add tests for unsafe ops. See src/flint/fmpz_abs_series-test.jl for some example tests you could pinch.

File: src/calcium/qqbar.jl [hunk 5]
```diff
@@ -0,0 +1,837 @@
...
+
+(a::CalciumQQBarField)() = qqbar()
+
+(a::CalciumQQBarField)(b::Int) = qqbar(b)
+
+(a::CalciumQQBarField)(b::fmpz) = qqbar(b)
+
+(a::CalciumQQBarField)(b::fmpq) = qqbar(b)
+
+(a::CalciumQQBarField)(b::qqbar) = b
```
Comment (wbhart) [inline src/calcium/qqbar.jl:836]: Any possibility to add a single line test for this under constructors or similar. eg. just test it has the right type.

File: src/calcium/qqbar.jl [hunk 6]
```diff
@@ -0,0 +1,837 @@
...
+function root_of_unity_as_args(a::qqbar)
+   p = Vector{Int}(undef, 1)
+   q = Vector{Int}(undef, 1)
+   if !Bool(ccall((:qqbar_is_root_of_unity, libcalcium), Cint, (Ptr{Int}, Ptr{Int}, Ref{qqbar}), p, q, a))
+      throw(DomainError(a, "value is not a root of unity"))
+   end
+   return (q[1], p[1])
+end
+
+function exp_pi_i(a::qqbar)
```
Comment (wbhart) [inline src/calcium/qqbar.jl:653]: Is there a test for this fn, or is codecov off its head?

File: src/calcium/qqbar.jl [hunk 7]
```diff
@@ -0,0 +1,837 @@
...
+isless(a::qqbar, b::fmpz) = isless(a, qqbar(b))
+isless(a::qqbar, b::fmpq) = isless(a, qqbar(b))
+isless(a::qqbar, b::Int) = isless(a, qqbar(b))
+isless(a::fmpq, b::qqbar) = isless(qqbar(a), b)
+isless(a::fmpz, b::qqbar) = isless(qqbar(a), b)
+isless(a::Int, b::qqbar) = isless(qqbar(a), b)
+
+# todo: name and export the following functions?
+
+function cmp_real(a::qqbar, b::qqbar)
```
Comment (wbhart) [inline src/calcium/qqbar.jl:423]: Are these cmp fns tested?

**Review by wbhart — CHANGES_REQUESTED** [review]

**Comment by fredrik-johansson** [comment]
> Someone will have to implement hash and expressify before this can be merged. But perhaps we can help out there. The other things you don't wish to add to this PR are fine by me.
> 
> I will take a look at the rest now.

Hash is trivial to do (I don't have a C function for this yet, but as a placeholder it will work well enough to hash the minimal polynomial). What is expected for expressify? It's not in the Nemo documentation.

The rand functions are a bit annoying because it will be quite a bit of code to reimplement them in Julia. It's too bad we don't just allow flint_rand_t to wrap the Julia RNGs.

**Comment by wbhart** [comment]
Yeah a simple hash like that is fine for now. Expressify is documented here:

https://nemocas.github.io/AbstractAlgebra.jl/latest/rings/

Ask @tthsqe12 for anything not documented.

Rand is indeed problematic. But it looks like we didn't get around to using the new rand model for arbs either, so we can omit this.

**Comment by fredrik-johansson** [comment]
Now a question about how to print qqbars. Which do you prefer?

* Current, compact form:

        -0.500000 + 0.866025*I (deg 2)

* Alternative using mathematical or Julia syntax:

        -0.500000 + 0.866025*i (deg 2)
        -0.500000 + 0.866025i (deg 2)
        -0.500000 + 0.866025im (deg 2)

* Descriptive form including the minimal polynomial:

        Root x ~ -0.500000 + 0.866025*I of x^2 + x + 1
        Root x ~ -0.500000 + 0.866025im of x^2 + x + 1

* Something you can copy and paste?

* Something else?

**Comment by wbhart** [comment]
I personally prefer the descriptive form, but let's see what others think.

**Comment by fredrik-johansson** [comment]
Printing now looks like this:

    julia> [qqbar(n) for n=-5:5]
    11-element Array{qqbar,1}:
     Root -5.00000 of x + 5
     Root -4.00000 of x + 4
     Root -3.00000 of x + 3
     Root -2.00000 of x + 2
     Root -1.00000 of x + 1
     Root 0 of x
     Root 1.00000 of x - 1
     Root 2.00000 of x - 2
     Root 3.00000 of x - 3
     Root 4.00000 of x - 4
     Root 5.00000 of x - 5

    julia> qqbar(355) // 113
    Root 3.14159 of 113x - 355

    julia> qqbar(1im)
    Root 1.00000*I of x^2 + 1

    julia> R, x = PolynomialRing(QQ, "x")
    (Univariate Polynomial Ring in x over Rational Field, x)

    julia> v = roots(x^5-x-1, CalciumQQBar)
    5-element Array{qqbar,1}:
     Root 1.16730 of x^5 - x - 1
     Root 0.181232 + 1.08395*I of x^5 - x - 1
     Root 0.181232 - 1.08395*I of x^5 - x - 1
     Root -0.764884 + 0.352472*I of x^5 - x - 1
     Root -0.764884 - 0.352472*I of x^5 - x - 1

    julia> sinpi(qqbar(1)//37)
    Root 0.0848059 of 68719476736x^36 - 635655159808x^34 + 2701534429184x^32 - 6992206757888x^30 + 12315818721280x^28 - 15625695002624x^26 + 14743599316992x^24 - 10531142369280x^22 + 5742196162560x^20 - 2392581734400x^18 + 757650882560x^16 - 180140769280x^14 + 31524634624x^12 - 3940579328x^10 + 336540160x^8 - 18356736x^6 + 573648x^4 - 8436x^2 + 37

I haven't put in code to print more digits for clustered roots, which will require adding a C helper function, but this should be fine for now.

**Comment by fredrik-johansson** [comment]
Regarding the untested comparison functions, I have not decided on an interface.

Julia seems to prefer having separate isless and ==/isequal functions for comparisons. This is silly performance-wise, because you have to do two computations instead of one, but it is reasonable interface-wise. With some redundancy, we could have all of the following:

(for performance):
cmp
cmp_real
cmp_imag
cmpabs_real
cmpabs_imag
cmpabs
cmp_root_order

(for julianness):
isequal/==
isequal_real
isequal_imag
isequal_abs
isless_real
isless_imag
isless_abs
isless_root_order

Is that OK?

**Comment by tthsqe12** [comment]
I wouldn't implement expressify at this point. This is code is perfect for now until we figure out what is going on with calcium.
```
function show(io::IO, x::qqbar)
   print(io, native_string(x))
end

needs_parentheses(x::qqbar) = true
```

**Comment by fredrik-johansson** [comment]
Ready for review.

Regarding inner constructors, what is the point? The only difference as far as I can tell is there will be more code duplication in order to ccall init and set the finalizer in each method.

File: src/calcium/qqbar.jl [hunk 8]
```diff
@@ -0,0 +1,1289 @@
...
+isless_abs_imag(a::qqbar, b::qqbar) = cmpabs_imag(a, b) < 0
+
+@doc Markdown.doc"""
+    isless_root_order(a::qqbar, b::qqbar)
+
+Compares the *a* and *b* in root sort order.
+"""
+isless_root_order(a::qqbar, b::qqbar) = cmp_root_order(a, b) < 0
+
+# todo: wrap qqbar_equal_fmpq_poly_val
```
Comment (wbhart) [inline src/calcium/qqbar.jl:655]: Did you intend to leave this todo?

File: src/calcium/qqbar.jl [hunk 9]
```diff
@@ -0,0 +1,1289 @@
...
+numbers *R*. The output array is sorted in the default sort order for
+algebraic numbers.
+"""
+function conjugates(a::qqbar)
+   deg = degree(a)
+   if deg == 1
+      return [a]
+   end
+   conjugates = qqbar_vec(deg)
+   ccall((:qqbar_conjugates, libcalcium), Nothing, (Ptr{qqbar_struct}, Ref{qqbar}), conjugates, a)
```
Comment (wbhart) [inline src/calcium/qqbar.jl:886]: Perhaps break some of these lines

File: src/calcium/qqbar.jl [hunk 10]
```diff
@@ -0,0 +1,1289 @@
...
+according to their multiplicity.
+"""
+function eigenvalues(A::fmpz_mat, R::CalciumQQBarField)
+   n = nrows(A)
+   !issquare(A) && throw(DomainError(A, "a square matrix is required"))
+   if n == 0
+      return Array{qqbar}(undef, 0)
+   end
+   roots = qqbar_vec(n)
+   ccall((:qqbar_eigenvalues_fmpz_mat, libcalcium), Nothing, (Ptr{qqbar_struct}, Ref{fmpz_mat}, Int), roots, A, 0)
```
Comment (wbhart) [inline src/calcium/qqbar.jl:907]: and here

File: src/calcium/qqbar.jl [hunk 11]
```diff
@@ -0,0 +1,1289 @@
...
+according to their multiplicity.
+"""
+function eigenvalues(A::fmpq_mat, R::CalciumQQBarField)
+   n = nrows(A)
+   !issquare(A) && throw(DomainError(A, "a square matrix is required"))
+   if n == 0
+      return Array{qqbar}(undef, 0)
+   end
+   roots = qqbar_vec(n)
+   ccall((:qqbar_eigenvalues_fmpq_mat, libcalcium), Nothing, (Ptr{qqbar_struct}, Ref{fmpq_mat}, Int), roots, A, 0)
```
Comment (wbhart) [inline src/calcium/qqbar.jl:928]: here

File: src/calcium/qqbar.jl [hunk 12]
```diff
@@ -0,0 +1,1289 @@
...
+    root_of_unity_as_args(a::qqbar)
+
+Return a pair of integers (*q*, *p*) such that the given *a* equals
+$e^{2 \pi i p / q}$. The denominator $q$ will be minimal, with
+$0 \le p < q$. Throws if *a* is not a root of unity.
+"""
+function root_of_unity_as_args(a::qqbar)
+   p = Vector{Int}(undef, 1)
+   q = Vector{Int}(undef, 1)
+   if !Bool(ccall((:qqbar_is_root_of_unity, libcalcium), Cint, (Ptr{Int}, Ptr{Int}, Ref{qqbar}), p, q, a))
```
Comment (wbhart) [inline src/calcium/qqbar.jl:985]: here

File: src/calcium/qqbar.jl [hunk 13]
```diff
@@ -0,0 +1,1289 @@
...
+@doc Markdown.doc"""
+    atanpi(a::qqbar)
+
+Return $\operatorname{atan}(a) / \pi$ as an algebraic number.
+Throws if this value is transcendental or undefined.
+"""
+function atanpi(a::qqbar)
+   p = Vector{Int}(undef, 1)
+   q = Vector{Int}(undef, 1)
+   if !Bool(ccall((:qqbar_atan_pi, libcalcium), Cint, (Ptr{Int}, Ptr{Int}, Ref{qqbar}), p, q, a))
```
Comment (wbhart) [inline src/calcium/qqbar.jl:1062]: etc

**Review by wbhart — CHANGES_REQUESTED** [review]

**Comment by wbhart** [comment]
Once the formatting issues are fixed, this is good to go from my point of view. Thanks!

**Comment by fredrik-johansson** [comment]
Fixed the long lines.

Yes, I intend to leave "wrap qqbar_equal_fmpq_poly_val" as a todo. It's a bit of an esoteric function; I might revisit it if I find an application where it's really important. However, this reminded me that I had forgotten to wrap the fmpz_poly and fmpq_poly evaluation functions. These don't do anything interesting in the current Calcium_jll but they will use a much faster algorithm in the next version, which also makes qqbar_equal_fmpq_poly_val less interesting. They are wrapped now.

**Review by wbhart — APPROVED** [review]

**Comment by fredrik-johansson** [comment]
@thofma This is ready to merge.

I will leave any additional features to future issues/PRs.

**Comment by thofma** [comment]
Awesome!

**Pull request merged**